File: ts/src/corpus.ts
import { CollisionError, EmbedderRequiredError, RefError } from "./errors.js";
import { nodeFromMarkdown, nodeToMarkdown } from "./frontmatter.js";
import { NodeId } from "./ids.js";
import type { Node } from "./node.js";
import type { Registry } from "./registry.js";
import { type SearchHit, SearchIndex } from "./search.js";
import { MEMBERSHIP } from "./shapes.js";
import { type Embedder, type SimilarHit, type Vector, VectorCache, VectorIndex } from "./similarity.js";
import {
  type ManifestEntry,
  type Snapshot,
  hashBytes,
  iterCorpusFiles,
  loadSnapshot,
  pathForNodeId,
  writeSnapshot,
} from "./snapshot.js";
import { Store } from "./store.js";
import { Index, type ResolvedEdge } from "./structural-index.js";

/** Rewrite every position in `node` that holds `oldId` to `newId` (in place). */
function rewriteRefs(node: Node, oldId: string, newId: string): void {
  for (const rel of node.relations) {
    if (rel.source === oldId) rel.source = newId;
    if (rel.target === oldId) rel.target = newId;
  }
  const mem = node.facets[MEMBERSHIP];
  if (mem !== undefined && mem !== null && typeof mem === "object") {
    const m = mem as Record<string, unknown>;
    const members = m.members;
    if (Array.isArray(members)) {
      m.members = members.map((x) => (x === oldId ? newId : x));
    } else if (members !== null && typeof members === "object") {
      const obj = members as Record<string, unknown>;
      for (const key of Object.keys(obj)) {
        if (obj[key] === oldId) obj[key] = newId;
      }
    }
    const edges = m.edges;
    if (Array.isArray(edges)) {
      for (const edge of edges) {
        if (edge !== null && typeof edge === "object") {
          const e = edge as Record<string, unknown>;
          if (e.source === oldId) e.source = newId;
          if (e.target === oldId) e.target = newId;
        }
      }
    }
  }
}

/** Coordinator over a `Store` + an in-memory `Index`. The primary kernel API. */
export class Corpus {
  readonly store: Store;
  readonly registry?: Registry;
  index!: Index;
  searchIndex!: SearchIndex;
  readonly embedder?: Embedder;
  readonly vectorCache?: VectorCache;
  vectorIndex?: VectorIndex;
  manifest: Map<string, ManifestEntry>;

  constructor(root: string, registry?: Registry, embedder?: Embedder) {
    this.store = new Store(root);
    this.registry = registry;
    this.embedder = embedder;
    this.vectorCache = embedder !== undefined ? new VectorCache(root) : undefined;
    this.manifest = new Map();
    const namespace = embedder !== undefined ? embedder.cacheNamespace : null;
    const snap = loadSnapshot(this.store.root, namespace);
    if (snap === null) this.fullRebuild();
    else this.reconcile(snap);
  }

  private relPath(nodeId: string): string {
    return pathForNodeId(nodeId);
  }

  private recordManifest(node: Node): void {
    const path = this.relPath(node.id);
    const data = Buffer.from(nodeToMarkdown(node), "utf-8");
    this.manifest.set(path, { path, sha256: hashBytes(data), uid: node.uid });
  }

  private fullRebuild(): void {
    const nodes: Node[] = [];
    const manifest = new Map<string, ManifestEntry>();
    for (const f of iterCorpusFiles(this.store.root)) {
      const node = nodeFromMarkdown(f.data.toString("utf-8"));
      nodes.push(node);
      manifest.set(f.path, { path: f.path, sha256: f.sha256, uid: node.uid });
    }
    this.index = Index.build(nodes);
    this.searchIndex = SearchIndex.build(nodes);
    if (this.embedder !== undefined) {
      this.vectorIndex = VectorIndex.build(nodes, this.embedder, this.vectorCache as VectorCache);
    } else {
      this.vectorIndex = undefined;
    }
    this.manifest = manifest;
  }

  private reconcile(snap: Snapshot): void {
    this.index = snap.index;
    this.searchIndex = snap.searchIndex;
    this.vectorIndex = snap.vectorIndex ?? undefined;
    const old = new Map<string, ManifestEntry>(snap.manifest.map((m) => [m.path, m]));
    const newManifest = new Map<string, ManifestEntry>();
    const changed: Array<{ path: string; sha256: string; node: Node }> = [];
    const drops: string[] = [];
    const current = new Set<string>();
    for (const f of iterCorpusFiles(this.store.root)) {
      current.add(f.path);
      const prev = old.get(f.path);
      if (prev !== undefined && prev.sha256 === f.sha256) {
        newManifest.set(f.path, prev); // unchanged: keep deserialized state, no parse
        continue;
      }
      if (prev !== undefined) drops.push(prev.uid);
      changed.push({ path: f.path, sha256: f.sha256, node: nodeFromMarkdown(f.data.toString("utf-8")) });
    }
    for (const [path, m] of old) {
      if (!current.has(path)) drops.push(m.uid); // deleted on disk
    }
    for (const uid of drops) {
      this.index.remove(uid);
      this.searchIndex.remove(uid);
      this.vectorIndex?.remove(uid);
    }
    for (const { path, sha256, node } of changed) {
      // Full build() collision contract: duplicate uid is rejected outright, then assertAddable.
      if (this.index.byUid.has(node.uid))
        throw new CollisionError(`duplicate uid ${JSON.stringify(node.uid)} in corpus`);
      this.index.assertAddable(node);
      const prepared =
        this.vectorIndex !== undefined
          ? this.vectorIndex.prepare(node, this.embedder as Embedder, this.vectorCache as VectorCache)
          : undefined;
      this.index.upsert(node);
      this.searchIndex.upsert(node);
      if (this.vectorIndex !== undefined && prepared !== undefined) this.vectorIndex.commit(node, prepared);
      newManifest.set(path, { path, sha256, uid: node.uid });
    }
    this.manifest = newManifest;
  }

  flushIndex(): void {
    const manifest = [...this.manifest.values()].sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0));
    writeSnapshot(this.store.root, manifest, this.index, this.searchIndex, this.vectorIndex);
  }

  private idFor(uid: string): string {
    const entry = this.index.byUid.get(uid);
    if (entry === undefined) throw new RefError(`uid ${JSON.stringify(uid)} not in index`);
    return entry.id;
  }

  private requireUid(ref: string): string {
    const uid = this.index.resolveUid(ref);
    if (uid === null) throw new RefError(`no node resolves ref ${JSON.stringify(ref)}`);
    return uid;
  }

  add(node: Node): Node {
    if (this.registry !== undefined) this.registry.validate(node);
    this.index.assertAddable(node);
    const prepared =
      this.vectorIndex !== undefined
        ? this.vectorIndex.prepare(node, this.embedder as Embedder, this.vectorCache as VectorCache)
        : undefined;
    this.store.writeFile(node);
    this.index.upsert(node);
    this.searchIndex.upsert(node);
    if (this.vectorIndex !== undefined && prepared !== undefined) {
      this.vectorIndex.commit(node, prepared);
    }
    return node;
  }

  get(ref: string): Node {
    return this.store.readFile(this.idFor(this.requireUid(ref)));
  }

  resolve(ref: string): Node {
    return this.get(ref);
  }

  delete(nodeId: string): void {
    const uid = this.index.idToUid.get(nodeId);
    if (uid === undefined) throw new RefError(`no live node at ${JSON.stringify(nodeId)}`);
    this.store.deleteFile(nodeId);
    this.index.remove(uid);
    this.searchIndex.remove(uid);
    this.vectorIndex?.remove(uid);
  }

  all(): Node[] {
    return this.store.allNodes();
  }

  outbound(ref: string): ResolvedEdge[] {
    return this.index.outboundEdges(this.requireUid(ref));
  }

  inbound(ref: string): ResolvedEdge[] {
    return this.index.inboundEdges(this.requireUid(ref));
  }

  dangling(): ResolvedEdge[] {
    return this.index.danglingEdges();
  }

  neighbors(ref: string): Node[] {
    const uid = this.requireUid(ref);
    const neighborUids = new Set<string>();
    for (const edge of this.index.outboundEdges(uid)) {
      if (edge.targetUid !== null) neighborUids.add(edge.targetUid);
    }
    for (const edge of this.index.inboundEdges(uid)) {
      if (edge.sourceUid !== null) neighborUids.add(edge.sourceUid);
    }
    neighborUids.delete(uid);
    return [...neighborUids].sort().map((u) => this.store.readFile(this.idFor(u)));
  }

  rename(oldId: string, newId: string): Node {
    // 1. oldId must be a LIVE id (not unknown, not merely deprecated); then collision-check newId.
    const uid = this.index.idToUid.get(oldId);
    if (uid === undefined) throw new RefError(`rename source ${JSON.stringify(oldId)} is not a live id`);
    if (this.index.resolveUid(newId) !== null) {
      throw new CollisionError(`target id ${JSON.stringify(newId)} already in use`);
    }

    // 2. Snapshot the referrer set BEFORE any index mutation (upsert rewrites inRefs).
    const referrerUids = new Set<string>();
    for (const inref of this.index.inRefs.get(oldId) ?? []) referrerUids.add(inref.sourceUid);

    // 3. Rewrite the renamed node itself (incl. its own oldId refs).
    const node = this.store.readFile(oldId);
    const oldPath = this.store.pathFor(oldId);
    node.id = newId;
    node.kind = NodeId.parse(newId).kind;
    if (!node.deprecatedIds.includes(oldId)) node.deprecatedIds.push(oldId);
    rewriteRefs(node, oldId, newId);

    // 4. Rewrite every OTHER referrer in memory.
    const referrers: Node[] = [];
    for (const referrerUid of referrerUids) {
      if (referrerUid === uid) continue;
      const referrer = this.store.readFile(this.idFor(referrerUid));
      rewriteRefs(referrer, oldId, newId);
      referrers.push(referrer);
    }

    // 5. Validate ALL writes before ANY write (fail-early, no partial rename).
    if (this.registry !== undefined) {
      this.registry.validate(node);
      for (const referrer of referrers) this.registry.validate(referrer);
    }

    // 5b. Prepare the renamed node's vector (fail before any disk write).
    const prepared =
      this.vectorIndex !== undefined
        ? this.vectorIndex.prepare(node, this.embedder as Embedder, this.vectorCache as VectorCache)
        : undefined;

    // 6. Commit: renamed node first (crash-atomic), then referrers. Each written once.
    const newPath = this.store.writeFile(node);
    if (oldPath !== newPath) this.store.deleteFile(oldId);
    this.index.upsert(node);
    for (const referrer of referrers) {
      this.store.writeFile(referrer);
      this.index.upsert(referrer);
    }

    this.searchIndex.upsert(node);
    if (this.vectorIndex !== undefined && prepared !== undefined) {
      this.vectorIndex.commit(node, prepared);
    }
    return node;
  }

  search(query: string, limit?: number): SearchHit[] {
    return this.searchIndex.search(query, limit);
  }

  similar(ref: string, k?: number): SimilarHit[] {
    if (this.vectorIndex === undefined) {
      throw new EmbedderRequiredError("similarity requires Corpus(root, registry?, embedder)");
    }
    return this.vectorIndex.similar(this.requireUid(ref), k);
  }

  queryVector(vec: Vector, k?: number): SimilarHit[] {
    if (this.vectorIndex === undefined) {
      throw new EmbedderRequiredError("similarity requires Corpus(root, registry?, embedder)");
    }
    return this.vectorIndex.queryVector(vec, k);
  }

  similarText(text: string, k?: number): SimilarHit[] {
    if (this.vectorIndex === undefined) {
      throw new EmbedderRequiredError("similarity requires Corpus(root, registry?, embedder)");
    }
    return this.vectorIndex.similarText(text, this.embedder as Embedder, k);
  }
}

File: python/src/nodes/kernel/corpus.py
from __future__ import annotations

from pathlib import Path

from nodes.kernel.errors import CollisionError, EmbedderRequiredError, RefError
from nodes.kernel.frontmatter import node_from_markdown
from nodes.kernel.ids import NodeId
from nodes.kernel.index import Index, ResolvedEdge
from nodes.kernel.node import Node
from nodes.kernel.registry import Registry
from nodes.kernel.search import SearchHit, SearchIndex
from nodes.kernel.shapes import MEMBERSHIP
from nodes.kernel.similarity import Embedder, SimilarHit, Vector, VectorCache, VectorIndex
from nodes.kernel.snapshot import (
    ManifestEntry,
    Snapshot,
    hash_bytes,
    iter_corpus_files,
    load_snapshot,
    write_snapshot,
)
from nodes.kernel.store import Store


def _rewrite_refs(node: Node, old: str, new: str) -> None:
    """Rewrite every position in `node` that holds `old` to `new` (in place)."""
    for rel in node.relations:
        if rel.source == old:
            rel.source = new
        if rel.target == old:
            rel.target = new
    mem = node.facets.get(MEMBERSHIP)
    if isinstance(mem, dict):
        members = mem.get("members")
        if isinstance(members, list):
            mem["members"] = [new if m == old else m for m in members]
        elif isinstance(members, dict):
            for key, val in list(members.items()):
                if val == old:
                    members[key] = new
        for edge in mem.get("edges", []) or []:
            if isinstance(edge, dict):
                if edge.get("source") == old:
                    edge["source"] = new
                if edge.get("target") == old:
                    edge["target"] = new


class Corpus:
    """Coordinator over a `Store` + an in-memory `Index`. The primary kernel API."""

    def __init__(self, root: Path, registry: Registry | None = None, embedder: Embedder | None = None) -> None:
        self.store = Store(root)
        self.registry = registry
        self.embedder = embedder
        self.vector_cache: VectorCache | None = VectorCache(root) if embedder is not None else None
        self.manifest: dict[str, ManifestEntry] = {}
        namespace = embedder.cache_namespace if embedder is not None else None
        snap = load_snapshot(self.store.root, namespace)
        if snap is None:
            self._full_rebuild()
        else:
            self._reconcile(snap)

    def _rel_path(self, node_id: str) -> str:
        return self.store.path_for(node_id).relative_to(self.store.root).as_posix()

    def _full_rebuild(self) -> None:
        nodes: list[Node] = []
        manifest: dict[str, ManifestEntry] = {}
        for f in iter_corpus_files(self.store.root):
            node = node_from_markdown(f.data.decode("utf-8"))
            nodes.append(node)
            manifest[f.path] = ManifestEntry(path=f.path, sha256=f.sha256, uid=node.uid)
        self.index = Index.build(nodes)
        self.search_index = SearchIndex.build(nodes)
        if self.embedder is not None:
            assert self.vector_cache is not None
            self.vector_index: VectorIndex | None = VectorIndex.build(nodes, self.embedder, self.vector_cache)
        else:
            self.vector_index = None
        self.manifest = manifest

    def _reconcile(self, snap: Snapshot) -> None:
        self.index = snap.index
        self.search_index = snap.search_index
        self.vector_index = snap.vector_index
        old = {m.path: m for m in snap.manifest}
        new_manifest: dict[str, ManifestEntry] = {}
        changed: list[tuple[str, str, Node]] = []
        drops: list[str] = []
        current: set[str] = set()
        for f in iter_corpus_files(self.store.root):
            current.add(f.path)
            prev = old.get(f.path)
            if prev is not None and prev.sha256 == f.sha256:
                new_manifest[f.path] = prev
                continue
            if prev is not None:
                drops.append(prev.uid)
            changed.append((f.path, f.sha256, node_from_markdown(f.data.decode("utf-8"))))
        for path, m in old.items():
            if path not in current:
                drops.append(m.uid)
        for uid in drops:
            self.index.remove(uid)
            self.search_index.remove(uid)
            if self.vector_index is not None:
                self.vector_index.remove(uid)
        for path, sha, node in changed:
            if node.uid in self.index.by_uid:
                raise CollisionError(f"duplicate uid {node.uid!r} in corpus")
            self.index.assert_addable(node)
            prepared = None
            if self.vector_index is not None:
                assert self.embedder is not None and self.vector_cache is not None
                prepared = self.vector_index.prepare(node, self.embedder, self.vector_cache)
            self.index.upsert(node)
            self.search_index.upsert(node)
            if self.vector_index is not None and prepared is not None:
                self.vector_index.commit(node, prepared)
            new_manifest[path] = ManifestEntry(path=path, sha256=sha, uid=node.uid)
        self.manifest = new_manifest

    def flush_index(self) -> None:
        manifest = sorted(self.manifest.values(), key=lambda m: m.path)
        write_snapshot(self.store.root, manifest, self.index, self.search_index, self.vector_index)

    def add(self, node: Node) -> Node:
        if self.registry is not None:
            self.registry.validate(node)
        self.index.assert_addable(node)
        prepared = None
        if self.vector_index is not None:
            assert self.embedder is not None and self.vector_cache is not None
            prepared = self.vector_index.prepare(node, self.embedder, self.vector_cache)
        path = self.store.write_file(node)
        self.index.upsert(node)
        self.search_index.upsert(node)
        if self.vector_index is not None and prepared is not None:
            self.vector_index.commit(node, prepared)
        rel_path = path.relative_to(self.store.root).as_posix()
        self.manifest[rel_path] = ManifestEntry(path=rel_path, sha256=hash_bytes(path.read_bytes()), uid=node.uid)
        return node

    def get(self, ref: str) -> Node:
        uid = self.index.resolve_uid(ref)
        if uid is None:
            raise RefError(f"no node resolves ref {ref!r}")
        return self.store.read_file(self.index.by_uid[uid].id)

    def resolve(self, ref: str) -> Node:
        return self.get(ref)

    def delete(self, node_id: str) -> None:
        uid = self.index.id_to_uid.get(node_id)
        if uid is None:
            raise RefError(f"no live node at {node_id!r}")
        self.store.delete_file(node_id)
        self.index.remove(uid)
        self.search_index.remove(uid)
        if self.vector_index is not None:
            self.vector_index.remove(uid)

    def all(self) -> list[Node]:
        return self.store.all_nodes()

    def _require_uid(self, ref: str) -> str:
        uid = self.index.resolve_uid(ref)
        if uid is None:
            raise RefError(f"no node resolves ref {ref!r}")
        return uid

    def outbound(self, ref: str) -> list[ResolvedEdge]:
        return self.index.outbound_edges(self._require_uid(ref))

    def inbound(self, ref: str) -> list[ResolvedEdge]:
        return self.index.inbound_edges(self._require_uid(ref))

    def dangling(self) -> list[ResolvedEdge]:
        return self.index.dangling_edges()

    def neighbors(self, ref: str) -> list[Node]:
        uid = self._require_uid(ref)
        neighbor_uids: set[str] = set()
        for edge in self.index.outbound_edges(uid):
            if edge.target_uid is not None:
                neighbor_uids.add(edge.target_uid)
        for edge in self.index.inbound_edges(uid):
            if edge.source_uid is not None:
                neighbor_uids.add(edge.source_uid)
        neighbor_uids.discard(uid)
        return [self.store.read_file(self.index.by_uid[u].id) for u in sorted(neighbor_uids)]

    def search(self, query: str, limit: int | None = None) -> list[SearchHit]:
        return self.search_index.search(query, limit)

    def similar(self, ref: str, k: int | None = None) -> list[SimilarHit]:
        if self.vector_index is None:
            raise EmbedderRequiredError("similarity requires Corpus(embedder=...)")
        return self.vector_index.similar(self._require_uid(ref), k)

    def query_vector(self, vec: Vector, k: int | None = None) -> list[SimilarHit]:
        if self.vector_index is None:
            raise EmbedderRequiredError("similarity requires Corpus(embedder=...)")
        return self.vector_index.query_vector(vec, k)

    def similar_text(self, text: str, k: int | None = None) -> list[SimilarHit]:
        if self.vector_index is None:
            raise EmbedderRequiredError("similarity requires Corpus(embedder=...)")
        assert self.embedder is not None
        return self.vector_index.similar_text(text, self.embedder, k)

    def rename(self, old_id: str, new_id: str) -> Node:
        if old_id not in self.index.id_to_uid:
            raise RefError(f"rename source {old_id!r} is not a live id")
        if self.index.resolve_uid(new_id) is not None:
            raise CollisionError(f"target id {new_id!r} already in use")

        uid = self.index.id_to_uid[old_id]
        referrer_uids = {ir.source_uid for ir in self.index.in_refs.get(old_id, [])}

        # --- prepare: rewrite every node that will change, in memory ---
        node = self.store.read_file(old_id)
        old_path = self.store.path_for(old_id)
        node.id = new_id
        node.kind = NodeId.parse(new_id).kind
        if old_id not in node.deprecated_ids:
            node.deprecated_ids.append(old_id)
        _rewrite_refs(node, old_id, new_id)

        referrers: list[Node] = []
        for referrer_uid in referrer_uids:
            if referrer_uid == uid:
                continue
            referrer = self.store.read_file(self.index.by_uid[referrer_uid].id)
            _rewrite_refs(referrer, old_id, new_id)
            referrers.append(referrer)

        # --- validate: ALL writes, before ANY write (fail-early, no partial rename) ---
        if self.registry is not None:
            self.registry.validate(node)
            for referrer in referrers:
                self.registry.validate(referrer)

        # --- prepare similarity vector (fail before any disk write) ---
        prepared = None
        if self.vector_index is not None:
            assert self.embedder is not None and self.vector_cache is not None
            prepared = self.vector_index.prepare(node, self.embedder, self.vector_cache)

        # --- commit: renamed node first (crash-atomic), then referrers ---
        new_path = self.store.write_file(node)
        if old_path != new_path:
            self.store.delete_file(old_id)
        self.index.upsert(node)
        for referrer in referrers:
            self.store.write_file(referrer)
            self.index.upsert(referrer)

        self.search_index.upsert(node)
        if self.vector_index is not None and prepared is not None:
            self.vector_index.commit(node, prepared)
        return node
